### PR TITLE
CLID-265,CLID-259,CLID-262: generates folders based on the filtered catalog digest

### PR DESCRIPTION
**Pull request by aguidirh**
# Description

This PR generates the folders based on the digest of the operator catalog filtered. Using the digest of the filtered catalog allows oc-mirror to identify if the operator catalog was already rebuilt previously, avoiding in this way rebuilding again unnecessarily.

With this approach will be also possible to filter a catalog based on the original catalog saved on the local cache.

The operator collector currently is a little complex and the refactorings to make it easier to understand and maintain will be handled in a separated PR.

Fixes [CLID-265](https://issues.redhat.com/browse/CLID-265)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

With the following ImageSetConfiguration:

```
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.15 
      packages:
       - name: aws-load-balancer-operator
       - name: 3scale-operator
       - name: node-observability-operator
```

I ran mirrorToDisk:

```
./bin/oc-mirror -c /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/alex-isc/isc.yaml file:///home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/clid-265 --alpha-ctlg-filter=true --v2
```

And diskToMirror

```
./bin/oc-mirror -c /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/alex-isc/isc.yaml --from file:///home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/clid-265 docker://localhost:6000 --alpha-ctlg-filter=true --v2 --dest-tls-verify=false
```

And MirrorToMirror

```
./bin/oc-mirror -c /home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/alex-isc/isc.yaml --workspace file:///home/aguidi/go/src/github.com/aguidirh/oc-mirror/alex-tests/clid-265 docker://localhost:6000 --alpha-ctlg-filter=true --v2 --dest-tls-verify=false
```
## Expected Outcome
All the flows above should pass successfully. 

NOTE: Delete, targetTag and targetCatalog were not tested yet.

### CODE REVIEW

**Review by sherine-k — COMMENTED** [review]
Hi @aguidirh 
A few comments and nits...
As you say in the description, the complexity is becoming high, but we can solve this gradually.
I think that the rebuild is happening twice, because [prepareM2D](https://github.com/openshift/oc-mirror/blob/3eed6e62054c67a6ba4ac17f300c703d17475d44/v2/internal/pkg/operator/common.go#L206-L212) is saving both to the cache and to the destination registry.
we need to discuss what to do in order not to rebuild twice, and not to push multiple times...

File: v2/internal/pkg/cli/executor.go [hunk 1]
```diff
@@ -1031,13 +1039,12 @@ func (o *ExecutorSchema) CollectAll(ctx context.Context) (v2alpha1.CollectorSche
 
 	// CLID-230 rebuild-catalogs
 	oImgs := oCollector.AllImages
-	if o.alphaCtlgFilter && (o.Opts.IsMirrorToDisk() || o.Opts.IsMirrorToMirror()) {
-		results, delImgs, err := o.ImageBuilder.RebuildCatalogs(ctx, oCollector)
+	if o.alphaCtlgFilter && (o.Opts.IsMirrorToDisk() || o.Opts.IsMirrorToMirror()) { //TODO ALEX ADD A CHECK TO SEE IF THE CATALOG WAS ALREADY REBUILT BEFORE, IF YES, SKIP, IT WILL AVOID REBUILDING CATALOGS ALREADY REBUILT PREVIOUSLY
```
Comment (lmzuccarelli) [inline v2/internal/pkg/cli/executor.go:1042]: @aguidirh - was the check for a previously built catalog to difficult that it could not be included in this PR ?

**Comment by sherine-k** [comment]
/retitle CLID-265,CLID-259,CLID-262: generates folders based on the filtered catalog digest

**Comment by lmzuccarelli** [comment]
Verified locally using the following isc

```
apiVersion: mirror.openshift.io/v2alpha1
kind: ImageSetConfiguration
mirror:
  operators:
    # - catalog: oci:///tmp/oci-catalog
    - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.14
      packages:
      - name: aws-load-balancer-operator
    - catalog: registry.redhat.io/redhat/certified-operator-index:v4.16
      targetCatalog: other/certified-operator-index
      packages:
      - name: gitlab-operator-kubernetes
    - catalog: registry.redhat.io/redhat/certified-operator-index:v4.15
      targetCatalog: other/certified-operator-index
      targetTag: v4.15.0
      packages:
      - name: gitlab-operator-kubernetes
    - catalog: registry.redhat.io/redhat/community-operator-index:v4.16
      targetTag: v4.16.0
      packages:
      - name: cert-utils-operator
```
